### PR TITLE
[ASR] Conformer global tokens in local attention

### PR DESCRIPTION
**Pull request by sam1373**
# What does this PR do ?

Adds Longformer-style attention for Conformer encoder (limited context + few tokens with full context attention). Unlike with just limited context attention, when using global tokens model needs to be fine-tuned with this approach to get good results. After fine-tuning results are improved particularly on long audio (even if it's much longer than training data). When using the default Fast Conformer configs with this attention, inference on audio above 1 hour is supported.

**Collection**: ASR

# Changelog 
- Add option to use global tokens
- Adds corresponding test
- Configs for training Conformer with this attention and mention in docs

# Usage

Add the following parameters to the Conformer encoder in config:

```  
    self_attention_model: rel_pos_local_attn # longformer-style attention (sliding window + global tokens)
    global_tokens: 1 # number of tokens that attend and are attended to by all tokens
    global_attn_separate: false # whether global tokens should use separate q,k,v layers
```

# Before your PR is "Ready for review"
**Pre checks**:
- [X] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [X] Did you write any new necessary tests?
- [X] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [X] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: examples/asr/conf/fastconformer/fast-conformer-long_ctc_bpe.yaml
Comment (titu1994) [inline examples/asr/conf/fastconformer/fast-conformer-long_ctc_bpe.yaml:11]: Add long-context

File: examples/asr/conf/fastconformer/fast-conformer-long_ctc_bpe.yaml
Comment (titu1994) [inline examples/asr/conf/fastconformer/fast-conformer-long_ctc_bpe.yaml:100]: Add a Note regarding the sections that are the primary difference compared to original model config

File: examples/asr/conf/fastconformer/fast-conformer-long_transducer_bpe.yaml
Comment (titu1994) [inline examples/asr/conf/fastconformer/fast-conformer-long_transducer_bpe.yaml:11]: Name update

File: examples/asr/conf/fastconformer/fast-conformer-long_transducer_bpe.yaml
Comment (titu1994) [inline examples/asr/conf/fastconformer/fast-conformer-long_transducer_bpe.yaml:110]: Same as above

File: nemo/collections/asr/parts/submodules/multi_head_attention.py
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/multi_head_attention.py:438]: Add shape info as comment here

File: nemo/collections/asr/parts/submodules/multi_head_attention.py
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/multi_head_attention.py:443]: Use better keyword variable name

File: nemo/collections/asr/parts/submodules/multi_head_attention.py
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/multi_head_attention.py:472]: How about make this protected instead of private? Put _ at the end of the method name

File: nemo/collections/asr/parts/submodules/multi_head_attention.py
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/multi_head_attention.py:512]: Same as above

**Review by titu1994 — COMMENTED** [review]
Overall it looks good, some minor changes commented. @VahidooX  for final review

File: docs/source/asr/models.rst
Comment (VahidooX) [inline docs/source/asr/models.rst:151]: Would you please add more detail on the benefits and limitations of this configurations compared to the regular one?
Also a reference to the Lonformer paper would help.

File: docs/source/asr/models.rst
Comment (VahidooX) [inline docs/source/asr/models.rst:153]: Do we have any docs on this local-attention with Longformer support?

File: examples/asr/conf/fastconformer/fast-conformer-long_ctc_bpe.yaml
Comment (VahidooX) [inline examples/asr/conf/fastconformer/fast-conformer-long_ctc_bpe.yaml:1]: Is it nice to have "-" and "_" in the name at the same time?

File: examples/asr/conf/fastconformer/fast-conformer-long_ctc_bpe.yaml
Comment (VahidooX) [inline examples/asr/conf/fastconformer/fast-conformer-long_ctc_bpe.yaml:9]: Let's move this part about the Longformer to the beginning. It would be also nice to give some hints why and when users may need to use this training. Maybe discuss about the global token feature?

File: examples/asr/conf/fastconformer/fast-conformer-long_ctc_bpe.yaml
Comment (VahidooX) [inline examples/asr/conf/fastconformer/fast-conformer-long_ctc_bpe.yaml:13]: @titu1994 does hydra have a feature that we create a config based on another config to avoid duplication of the conformer config every time? Just put the differences?

File: examples/asr/conf/fastconformer/fast-conformer-long_ctc_bpe.yaml
Comment (VahidooX) [inline examples/asr/conf/fastconformer/fast-conformer-long_ctc_bpe.yaml:4]: Have you trained/evaluated with this exact configuration?

File: nemo/collections/asr/parts/submodules/conformer_modules.py
Comment (VahidooX) [inline nemo/collections/asr/parts/submodules/conformer_modules.py:68]: Does enabling global_attn_separate helps?

File: nemo/collections/asr/modules/conformer_encoder.py
Comment (VahidooX) [inline nemo/collections/asr/modules/conformer_encoder.py:124]: What other values are valid for this parameter?

File: examples/asr/conf/fastconformer/fast-conformer-long_ctc_bpe.yaml
Comment (VahidooX) [inline examples/asr/conf/fastconformer/fast-conformer-long_ctc_bpe.yaml:102]: How global_tokens gets disabled? by setting to a number less than 1?

**Review by VahidooX — CHANGES_REQUESTED** [review]
LGTM! Left some minor comments.

File: examples/asr/conf/fastconformer/fast-conformer-long_ctc_bpe.yaml
Comment (titu1994) [reply examples/asr/conf/fastconformer/fast-conformer-long_ctc_bpe.yaml:13]: Yeah it's one of its main features. However, NeMo is supposed to have only flat configs 1:1 corresponding to some script. Such hierarchical configs do reduce duplication yes, but also complicate running code cause now you'll need to have full directories of configs just to be able to link everything together. 

All Nemo configs should be fully self contained

File: nemo/collections/asr/modules/conformer_encoder.py
Comment (sam1373) [reply nemo/collections/asr/modules/conformer_encoder.py]: Replaced this with a different parameter for global tokens spacing

File: examples/asr/conf/fastconformer/fast-conformer-long_ctc_bpe.yaml
Comment (sam1373) [reply examples/asr/conf/fastconformer/fast-conformer-long_ctc_bpe.yaml:106]: Yes, it is only on if there is at least 1

File: nemo/collections/asr/parts/submodules/conformer_modules.py
Comment (sam1373) [reply nemo/collections/asr/parts/submodules/conformer_modules.py:68]: In my tests results are pretty similar

File: docs/source/asr/models.rst
Comment (sam1373) [reply docs/source/asr/models.rst]: Added reference and more explanation

File: examples/asr/conf/fastconformer/fast-conformer-long_ctc_bpe.yaml
Comment (sam1373) [reply examples/asr/conf/fastconformer/fast-conformer-long_ctc_bpe.yaml:5]: Yes, it gets slightly worse results on ls than baseline

File: docs/source/asr/models.rst
Comment (sam1373) [reply docs/source/asr/models.rst:157]: I added link to paper, I don't think we really need the full explanation in the docs

File: examples/asr/conf/fastconformer/fast-conformer-long_ctc_bpe.yaml
Comment (VahidooX) [reply examples/asr/conf/fastconformer/fast-conformer-long_ctc_bpe.yaml:106]: Let's add a note here on how to disable global tokens.

File: examples/asr/conf/fastconformer/fast-conformer-long_ctc_bpe.yaml
Comment (sam1373) [reply examples/asr/conf/fastconformer/fast-conformer-long_ctc_bpe.yaml]: Moved part higher. Added mention of which parameters are changed, the parameters are pretty self-explanatory. The docs explains when training is needed

**Review by VahidooX — APPROVED** [review]
LGTM! Just left some minor comments. Let's wait for the second round of review from @titu1994.

File: docs/source/asr/models.rst
Comment (VahidooX) [reply docs/source/asr/models.rst:157]: I don't mean on how longformer works. I meant docs on how users may switch to longformer or use it for inference or training. Is there any place in the docs that we mention we support these feature?

Comment (sam1373) [reply docs/source/asr/models.rst:157]: Training is here, just switching to local context is mentioned here: https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/main/asr/results.html#inference-on-long-audio 
Switching to local+global isn't mentioned there because it won't work without additional training

Comment (VahidooX) [reply docs/source/asr/models.rst:157]: I see, it looks fine, but shouldn't we add how to do inference with models trained with global tokens here now that we are going to add training? Or the same command works for models trained with global tokens?

Comment (sam1373) [reply docs/source/asr/models.rst:157]: If the model is already trained with global tokens then you can do inference in the same way as any other model. Added links to the configs to that part as well

**Review by VahidooX — APPROVED** [review]
LGTM!

**Pull request merged**
